src/Compilers/CSharp/Test/Semantic/Semantics/CollectionLiteralTests.cs

```diff
@@ -0,0 +1,2966 @@
+                {
+                  // Code size        7 (0x7)
+                  .maxstack  1
+                  IL_0000:  ldc.i4.0
+                  IL_0001:  newarr     "int[]"
+                  IL_0006:  ret
+                }
+                """);
+            verifier.VerifyIL("Program.Create2", """
+                {
+                  // Code size       16 (0x10)
+                  .maxstack  4
+                  IL_0000:  ldc.i4.1
+                  IL_0001:  newarr     "object[]"
+                  IL_0006:  dup
+                  IL_0007:  ldc.i4.0
+                  IL_0008:  ldc.i4.0
+                  IL_0009:  newarr     "object"
+                  IL_000e:  stelem.ref
+                  IL_000f:  ret
+                }
+                """);
+            verifier.VerifyIL("Program.Create3", """
+                {
+                  // Code size       52 (0x34)
+                  .maxstack  7
+                  IL_0000:  ldc.i4.2
+                  IL_0001:  newarr     "object[]"
+                  IL_0006:  dup
+                  IL_0007:  ldc.i4.0
+                  IL_0008:  ldc.i4.1
+                  IL_0009:  newarr     "object"
+                  IL_000e:  dup
+                  IL_000f:  ldc.i4.0
+                  IL_0010:  ldc.i4.1
+                  IL_0011:  box        "int"
+                  IL_0016:  stelem.ref
+                  IL_0017:  stelem.ref
+                  IL_0018:  dup
+                  IL_0019:  ldc.i4.1
+                  IL_001a:  ldc.i4.2
+                  IL_001b:  newarr     "object"
+                  IL_0020:  dup
+                  IL_0021:  ldc.i4.0
+                  IL_0022:  ldc.i4.2
+                  IL_0023:  box        "int"
+                  IL_0028:  stelem.ref
+                  IL_0029:  dup
+                  IL_002a:  ldc.i4.1
+                  IL_002b:  ldc.i4.3
+                  IL_002c:  box        "int"
+                  IL_0031:  stelem.ref
+                  IL_0032:  stelem.ref
+                  IL_0033:  ret
+                }
+                """);
+        }
+
+        [Fact]
+        public void Array_03()
+        {
+            string source = """
+                class Program
+                {
+                    static void Main()
+                    {
+                        object o;
+                        o = (int[])[];
+                        o.Report();
+                        o = (long?[])[null, 2];
+                        o.Report();
+                    }
+                }
+                """;
+            CompileAndVerify(new[] { source, GetCollectionExtensions() }, expectedOutput: "[], [null, 2], ");
+        }
+
+        [Fact]
+        public void Array_04()
+        {
+            string source = """
+                class Program
+                {
+                    static void Main()
+                    {
+                        object[,] x = [];
+                        int[,] y = [null, 2];
```

Comment:
It feels like people may attempt to do `int[,] y = [[1, 2], [3, 4]]` (if I'm recalling/understanding what that notations means correctly). Consider also testing that.